### ISSUE TITLE
Helm deployment stuck in the init state

### ISSUE THREAD

**Issue by MaisenbacherD** [openebs/mayastor#1317]
**Describe the bug**
When deploying the helm chart for Mayastor v2.0.0 the pods are stuck in the init state.

**To Reproduce**

On a fresh k8s cluster I am deploying the helm chart:
```
kubectl label node worker-node01 openebs.io/engine=mayastor
kubectl label node worker-node02 openebs.io/engine=mayastor
kubectl label node worker-node03 openebs.io/engine=mayastor
helm repo add mayastor https://openebs.github.io/mayastor-extensions/
helm install mayastor mayastor/mayastor -n mayastor --create-namespace --set="etcd.replicaCount=1,etcd.persistence.storageClass=manual,etcd.livenessProbe.initialDelaySeconds=5,etcd.readinessProbe.initialDelaySeconds=5,loki-stack.loki.persistence.storageClassName=manual,obs.callhome.enabled=false" --version 2.0.0
```

The Mayastor pods are stuck in the init state:
```
vagrant@master-node:~/testfiles/mayastor$ kubectl get pods -n mayastor
NAME                                          READY   STATUS     RESTARTS   AGE
mayastor-agent-core-557c9c5cc8-rdwqz          0/2     Init:0/1   0          5m43s
mayastor-agent-ha-node-c5ltp                  0/1     Init:0/1   0          5m43s
mayastor-agent-ha-node-gdqn7                  0/1     Init:0/1   0          5m43s
mayastor-agent-ha-node-hnbch                  0/1     Init:0/1   0          5m43s
mayastor-api-rest-744d4c7f5-nb8gc             0/1     Init:0/2   0          5m43s
mayastor-csi-controller-8688994959-j2kqp      0/3     Init:0/1   0          5m43s
mayastor-csi-node-7qmzj                       2/2     Running    0          5m43s
mayastor-csi-node-m52z7                       2/2     Running    0          5m43s
mayastor-csi-node-qs75w                       2/2     Running    0          5m43s
mayastor-etcd-0                               1/1     Running    0          5m43s
mayastor-io-engine-24vgl                      0/2     Init:0/2   0          5m43s
mayastor-io-engine-5kv4f                      0/2     Init:0/2   0          5m43s
mayastor-io-engine-nd2jd                      0/2     Init:0/2   0          5m43s
mayastor-loki-0                               1/1     Running    0          5m43s
mayastor-operator-diskpool-7975bb4b85-j6kwr   0/1     Init:0/2   0          5m43s
mayastor-promtail-f6zmp                       1/1     Running    0          5m43s
mayastor-promtail-m7zt8                       1/1     Running    0          5m43s
mayastor-promtail-zbcv6                       1/1     Running    0          5m43s
```

**Expected behavior**
I would expect all mayastor pods to be running after about 3 min.

**OS info**
 - Distro: Debian 11
 - Kernel version 5.19.0
 - MayaStor v2.0.0

**Additional context**
The following helm install worked for me a few days ago:
```
helm install mayastor mayastor/mayastor -n mayastor --create-namespace --set="etcd.replicaCount=1,etcd.persistence.storageClass=manual,etcd.livenessProbe.initialDelaySeconds=5,etcd.readinessProbe.initialDelaySeconds=5,loki-stack.loki.persistence.storageClassName=manual" --version 2.0.0-rc.2
```

**Comment by Abhinandan-Purkait** [comment]
Hi @MaisenbacherD seems like the pods are waiting on the `init-containers`, most of the pods wait on the `mayastor-agent-core` to come up first. Can you please check the why the `mayastor-agent-core` init container is struck? Could be an issue with dns name resolution, as I can see etcd is already up.

**Comment by MaisenbacherD** [comment]
Hi @Abhinandan-Purkait,
Sure here the describe:
```
vagrant@master-node:~/testfiles/mayastor$ kubectl describe pod mayastor-agent-core-557c9c5cc8-rdwqz -n mayastor
Name:                 mayastor-agent-core-557c9c5cc8-rdwqz
Namespace:            mayastor
Priority:             2000000000
Priority Class Name:  system-cluster-critical
Service Account:      mayastor-service-account
Node:                 worker-node02/192.168.200.12
Start Time:           Thu, 09 Feb 2023 14:14:11 +0000
Labels:               app=agent-core
                      openebs.io/logging=true
                      openebs.io/release=mayastor
                      openebs.io/version=2.0.0
                      pod-template-hash=557c9c5cc8
Annotations:          <none>
Status:               Pending
IP:                   172.16.16.16
IPs:
  IP:           172.16.16.16
Controlled By:  ReplicaSet/mayastor-agent-core-557c9c5cc8
Init Containers:
  etcd-probe:
    Container ID:  containerd://103561668daee0f8307325343a2e3f72153a693b0c0ba2b8e27cd4c5ec0447d6
    Image:         busybox:latest
    Image ID:      docker.io/library/busybox@sha256:7b3ccabffc97de872a30dfd234fd972a66d247c8cfc69b0550f276481852627c
    Port:          <none>
    Host Port:     <none>
    Command:
      sh
      -c
      trap "exit 1" TERM; until nc -vzw 5 mayastor-etcd 2379; do date; echo "Waiting for etcd..."; sleep 1; done;
    State:          Running
      Started:      Thu, 09 Feb 2023 14:14:13 +0000
    Ready:          False
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-hrpq8 (ro)
Containers:
  agent-core:
    Container ID:
    Image:         docker.io/openebs/mayastor-agent-core:v2.0.0
    Image ID:
    Port:          50051/TCP
    Host Port:     0/TCP
    Args:
      -smayastor-etcd:2379
      --request-timeout=5s
      --cache-period=30s
      --grpc-server-addr=0.0.0.0:50051
    State:          Waiting
      Reason:       PodInitializing
    Ready:          False
    Restart Count:  0
    Limits:
      cpu:     1
      memory:  128Mi
    Requests:
      cpu:     500m
      memory:  32Mi
    Environment:
      RUST_LOG:          info
      MY_POD_NAME:       mayastor-agent-core-557c9c5cc8-rdwqz (v1:metadata.name)
      MY_POD_NAMESPACE:  mayastor (v1:metadata.namespace)
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-hrpq8 (ro)
  agent-ha-cluster:
    Container ID:
    Image:         docker.io/openebs/mayastor-agent-ha-cluster:v2.0.0
    Image ID:
    Port:          50052/TCP
    Host Port:     0/TCP
    Args:
      -g=0.0.0.0:50052
      --store=http://mayastor-etcd:2379
      --core-grpc=https://mayastor-agent-core:50051
    State:          Waiting
      Reason:       PodInitializing
    Ready:          False
    Restart Count:  0
    Limits:
      cpu:     100m
      memory:  64Mi
    Requests:
      cpu:     100m
      memory:  16Mi
    Environment:
      RUST_LOG:          info
      MY_POD_NAME:       mayastor-agent-core-557c9c5cc8-rdwqz (v1:metadata.name)
      MY_POD_NAMESPACE:  mayastor (v1:metadata.namespace)
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-hrpq8 (ro)
Conditions:
  Type              Status
  Initialized       False
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  kube-api-access-hrpq8:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Burstable
Node-Selectors:              kubernetes.io/arch=amd64
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 5s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 5s
Events:
  Type    Reason     Age   From               Message
  ----    ------     ----  ----               -------
  Normal  Scheduled  26m   default-scheduler  Successfully assigned mayastor/mayastor-agent-core-557c9c5cc8-rdwqz to worker-node02
  Normal  Pulling    26m   kubelet            Pulling image "busybox:latest"
  Normal  Pulled     26m   kubelet            Successfully pulled image "busybox:latest" in 758.912427ms
  Normal  Created    26m   kubelet            Created container etcd-probe
  Normal  Started    26m   kubelet            Started container etcd-probe
```

and here the log of the init container:

```
vagrant@master-node:~/testfiles/mayastor$ kubectl logs mayastor-agent-core-557c9c5cc8-rdwqz -c etcd-probe -n mayastor
nc: bad address 'mayastor-etcd'
Thu Feb  9 14:14:33 UTC 2023
Waiting for etcd...
...
```
Looks like `mayastor-etcd` is expected to be running but `mayastor-etcd-0` is actually running.

**Comment by Abhinandan-Purkait** [comment]
Yeah, its waiting on `mayastor-etcd`. Can you check `kubectl get endpoints -n mayastor`, just to check the svc is created properly?

**Comment by MaisenbacherD** [comment]
Yep, here the output for `kubectl get endpoints -n mayastor`:
```
NAME                             ENDPOINTS                             AGE
mayastor-agent-core                                                    97m
mayastor-api-rest                                                      97m
mayastor-etcd                    172.16.16.20:2380,172.16.16.20:2379   97m
mayastor-etcd-headless           172.16.16.20:2380,172.16.16.20:2379   97m
mayastor-loki                    172.16.16.17:3100                     97m
mayastor-loki-headless           172.16.16.17:3100                     97m
mayastor-metrics-exporter-pool                                         97m
```

**Comment by Abhinandan-Purkait** [comment]
Would you want try restarting the `coredns` pods and then the `mayastor-agent-core` and check if that changes anything? Suspecting an dns issue here, but cannot tell with certainty. Meanwhile I will try to setup a cluster with your distro and check on it, as we don't test on that distro, if that's fine?

**Comment by MaisenbacherD** [comment]
I restarted both `coredns` pods and the `mayastor-agent-core` respectively with the command 
```
kubectl get pod mayastor-agent-core-557c9c5cc8-rdwqz -n mayastor -o yaml | kubectl replace --force -f -
```

Nothing changed and the `mayastor-agent-core` is still waiting for the init container.

If it helps I can point you to my scripted cluster setup :)
What distro are you testing on?

**Comment by Abhinandan-Purkait** [comment]
Yeah, that would help. We only test on Ubuntu and NixOS. cc @tiagolobocastro @blaisedias Please, correct me if I am wrong here.

**Comment by MaisenbacherD** [comment]
This is the repo for my kluster setup https://github.com/MaisenbacherD/kubeadm-scripts/tree/wip-libvirsh-setup 

You would have to go through "Prerequisites" and "Setup" in the README.md :)
Note that this spins up 4 VMs - a master node and 3 worker nodes. You might need to adjust the resources depending on the machine you are planing to use: https://github.com/MaisenbacherD/kubeadm-scripts/blob/wip-libvirsh-setup/Vagrantfile#L50

Let me know if I can help you somehow or you run into issues :)

**Comment by datacore-tilangovan** [comment]
@MaisenbacherD Can you test whether the following helps you? something similar happened in microk8s, due to calico vxlan bug. Please try this and let us know. 

https://github.com/canonical/microk8s/issues/3695

`microk8s kubectl patch felixconfigurations default --patch '{"spec":{"featureDetectOverride":"ChecksumOffloadBroken=true"}}' --type=merge`

**Comment by MaisenbacherD** [comment]
Hi @datacore-tilangovan,
Thanks for that hint!
With that I realised that calico was not running at all because the manifest I fetched did not exist anymore.

After deploying calico I am able to helm install mayastor again.
The `kubectl patch felixconfigurations default --patch '{"spec":{"featureDetectOverride":"ChecksumOffloadBroken=true"}}' --type=merge` was not necessary for me :)


Thanks again!
